devito/passes/iet/engine.py

```diff
@@ -221,72 +221,110 @@ def abstract_efunc(efunc):
+            - DiscreteFunctions are renamed as "f0", "f1", ...
+            - Arrays are renamed as "a0", "a1", ...
+            - Objects are renamed as "o0", "o1", ...
     """
-    mapper = {}
-    counter = Counter()
-    for i in FindSymbols().visit(efunc):
-        if isinstance(i, AliasFunction):
-            base = i.name
-
-        elif i.is_DiscreteFunction:
-            base = 'f'
-
-            rkwargs = set(i.__rkwargs__) - {'initializer'}
-            kwargs = {k: getattr(i, k, None) for k in rkwargs}
-            kwargs['name'] = '%s%d' % (base, counter[base])
-            v = AliasFunction(aliased=i, **kwargs)
-
-            mapper.update({
-                i: v,
-                i.indexed: v.indexed,
-                i.dmap: v.dmap,
-                i._C_symbol: v._C_symbol,
-            })
-
-        elif i.is_Array:
-            # Use special names for special objects
-            if isinstance(i, Lock):
-                base = 'lock'
-            else:
-                base = 'a'
-
-            kwargs = {k: getattr(i, k, None) for k in i.__rkwargs__}
-            kwargs['name'] = '%s%d' % (base, counter[base])
-            v = type(i).__base__(**kwargs)
-
-            mapper.update({
-                i: v,
-                i.indexed: v.indexed,
-                i._C_symbol: v._C_symbol,
-            })
-
-        elif isinstance(i, CompositeObject):
-            base = 'o'
-
-            args = [getattr(i, k, None) for k in i.__rargs__]
-            args[0] = '%s%d' % (base, counter[base])
-            kwargs = {k: getattr(i, k, None) for k in i.__rkwargs__}
-            v = i.func(*args, **kwargs)
-
-            mapper[i] = v
+    functions = FindSymbols().visit(efunc)
 
-        else:
-            base = i.name
+    # Precedence rules make it possible to reconstruct objects that depend on
+    # higher priority objects
+    priority = [DiscreteFunction]
 
-            v = i
+    def key(i):
+        for p, cls in enumerate(priority, 1):
+            if isinstance(i, cls):
+                return p
+        return 0
+    functions = sorted(functions, key=key, reverse=True)
 
-        counter[base] += 1
+    # Build abstraction mappings
+    mapper = {}
+    counter = Counter()
+    for i in functions:
+        abstract_object(i, mapper, counter)
 
     efunc = Uxreplace(mapper).visit(efunc)
     efunc = efunc._rebuild(name='foo')
 
     return efunc
 
 
+@singledispatch
+def abstract_object(i, mapper, counter, cls=None):
+    """
+    Singledispatch-based implementation of object abstraction.
+
+    Singledispatch allows foreign modules to specify their own rules for
+    object abstraction.
+    """
+    return
+
+
+@abstract_object.register(DiscreteFunction)
+def _(i, mapper, counter):
+    base = 'f'
+
+    kwargs = {k: getattr(i, k, None) for k in i.__rkwargs__}
+    kwargs['name'] = '%s%d' % (base, counter[base])
+    kwargs['initializer'] = None
+    kwargs['alias'] = True
+    v = i._rebuild(**kwargs)
```

Comment:
Isn't this a bit redundant/overkill? `__rkwargs__` will be processed by `_rebuild` so there is no need to do it here, `name/initializer/alias` should be the only ones needed

Reply:
jeeeeezzz ! True!